### PR TITLE
feat: Instance create

### LINKED ISSUE #4417: Create service account with credentials in instance setup
Description

To use the instance setup in automation, a service account has to be created with IAM_OWNER permissions and credentials, so that for example terraform can fully automated create an instance and all the other necessary resources on top.

**Acceptance criteria**
- [x] Create service account in instance setup
- [x] Give service account permissions on instance
- [x] Return credentials from instance setup for direct use

### PR DESCRIPTION
**Pull request by stebenz**
closes #4417

### CODE REVIEW

File: internal/command/instance.go [hunk 1]
```diff
@@ -366,14 +378,40 @@ func (c *Commands) SetUpInstance(ctx context.Context, setup *InstanceSetup) (str
...
-		return "", nil, err
+		return "", "", nil, nil, err
 	}
-	return instanceID, &domain.ObjectDetails{
+
+	var pat string
+	var machineKey []byte
+	if setup.Org.Machine != nil {
+		if setup.Org.Machine.Pat {
+			_, token, err := c.AddPersonalAccessToken(ctx, userID, orgID, setup.Org.Machine.PatExpirationDate, setup.Org.Machine.PatScopes, domain.UserTypeMachine)
```
Comment (livio-a) [inline internal/command/instance.go:393]: this command is no prepareFunction and therefor won't work in this context

File: internal/command/instance.go [hunk 2]
```diff
@@ -366,14 +378,40 @@ func (c *Commands) SetUpInstance(ctx context.Context, setup *InstanceSetup) (str
...
+			if err != nil {
+				return "", "", nil, nil, err
+			}
+			pat = token
+		}
+		if setup.Org.Machine.MachineKey {
+			key, err := c.AddUserMachineKey(ctx, &domain.MachineKey{
```
Comment (livio-a) [inline internal/command/instance.go:400]: this command is no prepareFunction and therefor won't work in this context

File: internal/command/org.go [hunk 3]
```diff
@@ -44,23 +47,55 @@ func (c *Commands) setUpOrgWithIDs(ctx context.Context, o *OrgSetup, orgID, user
...
 	if err != nil {
-		return "", nil, err
+		return "", "", nil, nil, err
+	}
+
+	var pat string
+	var machineKey []byte
+	if o.Machine != nil {
+		if o.Machine.Pat {
+			_, token, err := c.AddPersonalAccessToken(ctx, userID, orgID, o.Machine.PatExpirationDate, o.Machine.PatScopes, domain.UserTypeMachine)
```
Comment (livio-a) [inline internal/command/org.go:77]: this command is no prepareFunction and therefor won't work in this context

File: internal/command/org.go [hunk 4]
```diff
@@ -44,23 +47,55 @@ func (c *Commands) setUpOrgWithIDs(ctx context.Context, o *OrgSetup, orgID, user
...
+			if err != nil {
+				return "", "", nil, nil, err
+			}
+			pat = token
+		}
+		if o.Machine.MachineKey {
+			key, err := c.AddUserMachineKey(ctx, &domain.MachineKey{
```
Comment (livio-a) [inline internal/command/org.go:84]: this command is no prepareFunction and therefor won't work in this context

File: internal/command/user_machine.go [hunk 5]
```diff
@@ -2,108 +2,166 @@ package command
...
+			if isUserStateExists(writeModel.UserState) {
+				return nil, caos_errs.ThrowPreconditionFailed(nil, "COMMAND-k2una", "Errors.User.AlreadyExisting")
+			}
+			domainPolicy, err := domainPolicyWriteModel(ctx, filter)
+			if err != nil {
+				return nil, caos_errs.ThrowPreconditionFailed(err, "COMMAND-3M9fs", "Errors.Org.DomainPolicy.NotFound")
+			}
+			if !domainPolicy.UserLoginMustBeDomain {
+				return nil, caos_errs.ThrowPreconditionFailed(nil, "COMMAND-6M0dd", "Errors.User.Invalid")
+			}
```
Comment (livio-a) [inline internal/command/user_machine.go:44]: @hifabienne didn't we change that or did we only want to?

**Review by livio-a — CHANGES_REQUESTED** [review]

File: internal/command/user_machine.go (hunk 5 above)
Comment (hifabienne) [reply internal/command/user_machine.go:44]: Yes this should be removed

**Comment by hifabienne** [comment]
@stebenz @livio-a what is the state here?

**Comment by fforootd** [comment]
Just a minor thing, but can we name this PR according to semrel. Thanks

**Comment by livio-a** [comment]
> Just a minor thing, but can we name this PR according to semrel. Thanks

he old sem rel check app has been discontinued, i've therefore just installed an alternative and set up the check as required again

File: internal/api/grpc/system/instance_converter.go [hunk 6]
```diff
@@ -3,17 +3,109 @@ package system
...
+		defaultInstance.Org.Machine = nil
+	}
+
+	if lang := language.Make(req.DefaultLanguage); lang != language.Und {
+		defaultInstance.DefaultLanguage = lang
+	}
+
+	return &defaultInstance
+}
+func createInstancePbToAddHuman(user *system_pb.CreateInstanceRequest_Human, userLoginMustBeDomain bool, org, externalDomain string) *command.AddHuman {
```
Comment (livio-a) [inline internal/api/grpc/system/instance_converter.go:47]: please use the defaultHuman and add a line above

File: internal/api/grpc/system/instance_converter.go [hunk 7]
```diff
@@ -3,17 +3,109 @@ package system
...
+func createInstancePbToAddMachine(user *system_pb.CreateInstanceRequest_Machine) *command.AddMachine {
+	machine := &command.AddMachine{
+		Machine: &command.Machine{},
+	}
+	if user.UserName != "" {
+		machine.Machine.Username = user.UserName
+	}
+	if user.Name != "" {
+		machine.Machine.Name = user.Name
+	}
```
Comment (livio-a) [inline internal/api/grpc/system/instance_converter.go:91]: these ifs are unnecessary

**Review by livio-a — CHANGES_REQUESTED** [review]

File: internal/command/instance.go [hunk 8]
```diff
@@ -370,16 +381,49 @@ func (c *Commands) SetUpInstance(ctx context.Context, setup *InstanceSetup) (str
...
+		}
+		if setup.Org.Machine.MachineKey {
+			machineKey = NewMachineKey(orgID, userID, setup.Org.Machine.MachineKeyExpirationDate, setup.Org.Machine.MachineKeyType)
+			machineKey.KeyID, err = c.idGenerator.Next()
+			if err != nil {
+				return "", "", nil, nil, err
+			}
+			validations = append(validations, prepareAddUserMachineKey(machineKey, c.machineKeySize))
+		}
+	}
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/instance.go]: good point, moved it to the machine add, don't know why I placed is separately

File: internal/command/user_personal_access_token.go [hunk 9]
```diff
@@ -2,91 +2,190 @@ package command
...
+					pat.Scopes,
+				),
+			}, nil
+		}, nil
 	}
+}
 
-	pushedEvents, err := c.eventstore.Push(ctx,
-		user.NewPersonalAccessTokenRemovedEvent(ctx, UserAggregateFromWriteModel(&tokenWriteModel.WriteModel), tokenID))
+func (c *Commands) RemovePersonalAccessToken(ctx context.Context, pat *PersonalAccessToken) (*domain.ObjectDetails, error) {
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/user_personal_access_token.go:135]: to re-use the content check which you have to do in every action with the PAT, so to check if the 3 id strings are not empty, can change it if necessary

File: internal/command/user_machine_key.go [hunk 10]
```diff
@@ -2,108 +2,195 @@ package command
...
+					machineKey.PublicKey,
+				),
+			}, nil
+		}, nil
 	}
+}
 
-	pushedEvents, err := c.eventstore.Push(ctx,
-		user.NewMachineKeyRemovedEvent(ctx, UserAggregateFromWriteModel(&keyWriteModel.WriteModel), keyID))
+func (c *Commands) RemoveUserMachineKey(ctx context.Context, machineKey *MachineKey) (*domain.ObjectDetails, error) {
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/user_machine_key.go:152]: to re-use the content check which you have to do in every action with the machine key, so to check if the 3 id strings are not empty, can change it if necessary

File: internal/command/user.go [hunk 11]
```diff
@@ -396,6 +396,17 @@ func (c *Commands) checkUserExists(ctx context.Context, userID, resourceOwner st
 	return nil
 }
 
+func checkUserExists(ctx context.Context, filter preparation.FilterToQueryReducer, userID, resourceOwner string) error {
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/user.go]: remove the function and use ExistsUser which is also used in the member logic

File: internal/command/user_machine.go [hunk 12]
```diff
@@ -11,6 +12,16 @@ import (
...
 )
 
+type AddMachine struct {
+	Machine                  *Machine
+	Pat                      bool
+	PatExpirationDate        time.Time
+	PatScopes                []string
+	MachineKey               bool
+	MachineKeyType           domain.AuthNKeyType
+	MachineKeyExpirationDate time.Time
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/user_machine.go]: refactored

File: internal/api/grpc/system/instance_converter.go (hunk 6 above)
Comment (stebenz) [reply internal/api/grpc/system/instance_converter.go]: defaulthuman and defaultmachine are now used in the functions

File: internal/api/grpc/system/instance_converter.go (hunk 7 above)
Comment (stebenz) [reply internal/api/grpc/system/instance_converter.go]: all unnecessary ifs are removed

File: internal/api/grpc/system/instance_converter.go [hunk 13]
```diff
@@ -3,17 +3,107 @@ package system
...
+
+	if lang := language.Make(req.DefaultLanguage); lang != language.Und {
+		defaultInstance.DefaultLanguage = lang
+	}
+
+	return &defaultInstance
+}
+
+func createInstancePbToAddHuman(user *system_pb.CreateInstanceRequest_Human, defaultHuman *command.AddHuman, userLoginMustBeDomain bool, org, externalDomain string) *command.AddHuman {
+	if defaultHuman == nil {
```
Comment (livio-a) [inline internal/api/grpc/system/instance_converter.go:49]: As the AddHuman is changed to a pointer, every overwrite will change the default
Same obviously for machine

File: internal/command/instance.go [hunk 14]
```diff
@@ -372,14 +402,24 @@ func (c *Commands) SetUpInstance(ctx context.Context, setup *InstanceSetup) (str
...
+
+	var token string
+	var key []byte
+	if pat != nil {
+		token = pat.Token
 	}
-	return instanceID, &domain.ObjectDetails{
+	if machineKey != nil {
+		key = machineKey.PrivateKey
```
Comment (livio-a) [inline internal/command/instance.go:419]: The returned key must be computed with the detail function

File: internal/command/org.go [hunk 15]
```diff
@@ -44,23 +46,59 @@ func (c *Commands) setUpOrgWithIDs(ctx context.Context, o *OrgSetup, orgID, user
...
 	}
-	return userID, &domain.ObjectDetails{
+
+	var token string
+	var key []byte
+	if pat != nil {
+		token = pat.Token
+	}
+	if machineKey != nil {
+		key = machineKey.PrivateKey
```
Comment (livio-a) [inline internal/command/org.go:98]: ...detail function

**Review by livio-a — CHANGES_REQUESTED** [review]

File: cmd/setup/steps.yaml [hunk 16]
```diff
@@ -22,3 +22,9 @@ FirstInstance:
         Verified:
       Password: Password1!
       PasswordChangeRequired: true
+    Machine:
+      Machine:
+        Username:
+        Name:
+      MachineKey:
+        Type:
```
Comment (livio-a) [inline cmd/setup/steps.yaml:30]: please add an empty `MachineKeyPath` field to enable overwrite by env var

File: internal/api/grpc/system/instance.go [hunk 17]
```diff
@@ -62,6 +62,23 @@ func (s *Server) UpdateInstance(ctx context.Context, req *system_pb.UpdateInstan
...
 
+func (s *Server) CreateInstance(ctx context.Context, req *system_pb.CreateInstanceRequest) (*system_pb.CreateInstanceResponse, error) {
+	id, pat, key, details, err := s.command.SetUpInstance(ctx, CreateInstancePbToSetupInstance(req, s.defaultInstance, s.externalDomain))
+	if err != nil {
+		return nil, err
+	}
+	machineKey, err := key.Detail()
+	if err != nil {
+		return nil, err
+	}
```
Comment (livio-a) [inline internal/api/grpc/system/instance.go:73]: please check for a nil pointer on `key`

**Review by livio-a — CHANGES_REQUESTED** [review]

File: internal/api/grpc/system/instance_converter.go (hunk 7 above)
Comment (livio-a) [reply internal/api/grpc/system/instance_converter.go]: with the refactored code, the ifs would now make sense 😄

**Review by livio-a — APPROVED** [review]

**Pull request merged**